More robustly deal with relaxed bounds and improve their diagnostics

fmease opened the pull request:
Scaffolding for https://github.com/rust-lang/rust/issues/135229 (CC https://github.com/rust-lang/rust/pull/135331)

Fixes https://github.com/rust-lang/rust/issues/136944 (6th commit).
Fixes https://github.com/rust-lang/rust/issues/142718 (8th commit).

fmease committed:
HIR ty lowering: Simplify signature of `lower_poly_trait_ref`

fmease committed:
HIR ty lowering: Validate relaxed bounds in trait object types

Only relevant to the internal feature `more_maybe_bounds`.

fmease committed:
Update comment about `where Ty:`

We no longer move trait predicates where the self ty is a ty param to
"the bounds of a ty param".

fmease committed:
AST lowering: More robustly deal with relaxed bounds

rust-log-analyzer commented:
The job **`mingw-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15748165047/job/44388308387) [(plain)](https://github.com/rust-lang/rust/commit/03d01fa189bc4acd2920b5ad6a40f26d9f61c555/checks/44388308387/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

error[E0425]: cannot find function `feature_err` in this scope
    --> compiler/rustc_ast_lowering/src/lib.rs:2065:29
     |
2065 |         let err = |message| feature_err(&self.tcx.sess, sym::more_maybe_bounds, span, message);
     |                             ^^^^^^^^^^^ not found in this scope
     |
help: consider importing this function
     |
41   + use rustc_session::parse::feature_err;
```

</details>

fmease committed:
Reword diagnostic about relaxing non-`Sized` bound

* The phrasing "only does something for" made sense back when this
  diagnostic was a (hard) warning. Now however, it's simply a hard
  error and thus completely rules out "doing something".
* The primary message was way too long
* The new wording more closely mirrors the wording we use for applying
  other bound modifiers (like `const` and `async`) to incompatible
  traits.
* "all other traits are not bound by default" is no longer accurate
  under Sized Hierarchy. E.g., traits and assoc tys are (currently)
  bounded by `MetaSized` by default but can't be relaxed using
  `?MetaSized` (instead, you relax it by adding `PointeeSized`).
* I've decided against adding any diagnositic notes or suggestions
  for now like "trait `Trait` can't be relaxed as it's not bound by
  default" which would be incorrect for `MetaSized` and assoc tys
  as mentioned above) or "consider changing `?MetaSized` to
  `PointeeSized`" as the Sized Hierarchy impl is still WIP)

fmease committed:
Reword diagnostics about relaxed bounds in invalid contexts

fmease committed:
HIR ty lowering: Validate `PointeeSized` bounds

fmease commented:
@bors2 try rust-timer queue

fmease commented:
@bors2 try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued bfa10de6407128926b18c60d51c9f19df73e336c with parent 9c4ff566babe632af5e30281a822d1ae9972873b, future [comparison URL](https://perf.rust-lang.org/compare.html?start=9c4ff566babe632af5e30281a822d1ae9972873b&end=bfa10de6407128926b18c60d51c9f19df73e336c).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.5 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (bfa10de6407128926b18c60d51c9f19df73e336c): [comparison URL](https://perf.rust-lang.org/compare.html?start=9c4ff566babe632af5e30281a822d1ae9972873b&end=bfa10de6407128926b18c60d51c9f19df73e336c&stat=instructions:u).

### Overall result: ❌ regressions - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=9c4ff566babe632af5e30281a822d1ae9972873b&end=bfa10de6407128926b18c60d51c9f19df73e336c&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.0% | [3.0%, 3.0%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.0% | [3.0%, 3.0%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9c4ff566babe632af5e30281a822d1ae9972873b&end=bfa10de6407128926b18c60d51c9f19df73e336c&stat=max-rss)
<details>
<summary>Results (secondary -2.7%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.4%  | [2.4%, 2.4%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.5% | [-5.8%, -1.8%] | 6     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9c4ff566babe632af5e30281a822d1ae9972873b&end=bfa10de6407128926b18c60d51c9f19df73e336c&stat=cycles:u)
<details>
<summary>Results (primary 2.9%, secondary 4.4%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.9% | [2.9%, 2.9%] | 1     |
| Regressions ❌ <br /> (secondary)  | 4.4% | [2.8%, 7.6%] | 10    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.9% | [2.9%, 2.9%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=9c4ff566babe632af5e30281a822d1ae9972873b&end=bfa10de6407128926b18c60d51c9f19df73e336c&stat=size:linked_artifact)
<details>
<summary>Results (primary 1.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1% | [1.1%, 1.1%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.1% | [1.1%, 1.1%] | 1     |
</details>

**Bootstrap**: 692.606s -> 692.191s (-0.06%)
**Artifact size**: 372.01 MiB -> 371.99 MiB (-0.01%)

bors commented:
:umbrella: The latest upstream changes (presumably #143091) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease commented:
> https://github.com/rust-lang/rust/pull/142693#issuecomment-2993257784

Looks spurious?

rust-timer commented:
Queued 6893bb554373e6438fbf77486be0a58c34e372d4 with parent 7ba34c704529e7fcab80130c3fe40efe415d61b5, future [comparison URL](https://perf.rust-lang.org/compare.html?start=7ba34c704529e7fcab80130c3fe40efe415d61b5&end=6893bb554373e6438fbf77486be0a58c34e372d4).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

fmease commented on compiler/rustc_ast_lowering/src/lib.rs:
```diff
@@ -2011,19 +2034,77 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
+        span: Span,
+        rbp: RelaxedBoundPolicy<'_>,
+    ) {
+        let err = |message| feature_err(&self.tcx.sess, sym::more_maybe_bounds, span, message);
```
I'm not super fond of the fact that we currently advertise the *internal* feature `more_maybe_bounds`. I'd rather get rid of the feature suggestion similar to the internal feature `negative_bounds`. LMKWYT

compiler-errors replied:
I prefer as well that we remove it, though that can be a follow-up.

fmease commented on compiler/rustc_ast_lowering/src/lib.rs:
```diff
@@ -2011,19 +2034,77 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
+        match rbp {
+            RelaxedBoundPolicy::Allowed => return,
+            RelaxedBoundPolicy::AllowedIfOnTyParam(id, params) => {
+                if let Some(res) = self.resolver.get_partial_res(id).and_then(|r| r.full_res())
+                    && let Res::Def(DefKind::TyParam, def_id) = res
+                    && params.iter().any(|p| def_id == self.local_def_id(p.id).to_def_id())
+                {
+                    return;
+                }
```
The old approach in `lower_generics` cached whether the bounded ty was a "close ty param" which we no longer do.

However if you think about it, the caching on master doesn't make any sense in practice for real user-written code: It did avoid recomputation in `where T: ?Sized + ?Sized + ?Sized` (nobody writes multiple relaxed bounds since they're later rejected during HIR ty lowering) but it didn't avoid the recalculation in `where T: ?Sized, U: ?Sized, V: ?Sized` which is what users actually write (to be clear, the new code also recomputes this, it just throws out the unnecessary caching logic).

fmease commented on compiler/rustc_ast_lowering/src/lib.rs:
```diff
@@ -2011,19 +2034,77 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
+                        diag.emit();
+                        return;
+                    }
+                    RelaxedBoundForbiddenReason::LateBoundVarsInScope => {}
```
We could emit a custom diagnostic for this but it's simply not worth it.

fmease commented on compiler/rustc_ast_lowering/src/lib.rs:
```diff
@@ -2103,7 +2103,11 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
-        err("`?Trait` bounds are only permitted at the point where a type parameter is declared")
+        err("this relaxed bound is not permitted here")
+            .with_note(
+                "in this context, relaxed bounds are only allowed on \
```
"in this context" because in other contexts like in assoc tys, relaxed bounds are also allowed in other places (in that case: assoc ty item bounds).

fmease commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:
```diff
@@ -758,13 +758,37 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
+            }
+            rustc_ast::BoundPolarity::Positive => (ty::PredicatePolarity::Positive, bounds),
+            rustc_ast::BoundPolarity::Negative(_) => (ty::PredicatePolarity::Negative, bounds),
+            // FIXME(fmease): This is super hacky! `Option<Polarity>` is also sad.
```
(Yeah I should remove that FIXME, I forgot) I don't really like this structuring, very open to suggestions to address this FIXME.

**Bigger picture**: As you can see a few lines further down we used to duplicate the `lower_assoc_item_constraint` to catch bad bounds like `?Sized<Undefined = ()>`. This is obv annoying and not scalable (e.g., for `PointeeSized` we also need to validate the bound constness which is a lot more code I don't want to dupe / extract into a method (hmmm...)). So the new approach is to do as much normal lowering for these "fake bounds" but send everything into the drain (`&mut Vec::new()`).

fmease commented:
r? compiler-errors or reassign if you're busy or not interested

fmease commented:
Oh, a few tests are failing since the last rebase, I need to fix them up, my bad. Gimme a moment

fmease commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/dyn_compatibility.rs:
```diff
@@ -70,13 +69,6 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             let guar = self.report_trait_object_addition_traits(&regular_traits);
             return Ty::new_error(tcx, guar);
         }
-        // We don't support `PointeeSized` principals
```
My PR predates #143104.

This commit basically reverts this logic because it's unreachable due to us filtering out `PointeeSized` inside `lower_poly_trait_ref`. The diagnostic is definitely worse, we now say `at least one trait is required for an object type` instead of `` `PointeeSized` cannot be used with trait objects ``.

I guess I should address the diagnostic regression since it only makes sense if you know the lowering. This ties in with https://github.com/rust-lang/rust/pull/142693#discussion_r2173559498.

fmease commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs:
```diff
@@ -209,9 +209,22 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
+            //        reject relaxed bounds outright during AST lowering.
+            //        However, this can easily get out of sync! Ideally, we would perform this step
+            //        where we are guaranteed to catch *all* bounds like in
+            //        `Self::lower_poly_trait_ref`. List of concrete issues:
```
(We likely can't actually do it in `lower_poly_trait_ref` because we need access to all the neighboring HIR bounds)

fmease commented on compiler/rustc_ast_lowering/src/lib.rs:
```diff
@@ -2011,19 +2034,77 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
+        match rbp {
+            RelaxedBoundPolicy::Allowed => return,
+            RelaxedBoundPolicy::AllowedIfOnTyParam(id, params) => {
+                if let Some(res) = self.resolver.get_partial_res(id).and_then(|r| r.full_res())
+                    && let Res::Def(DefKind::TyParam, def_id) = res
+                    && params.iter().any(|p| def_id == self.local_def_id(p.id).to_def_id())
```
(If you're curious, we can't simply replace this with something that's based on `rustc_middle::ty` and located in HIR ty lowering while preserving behavior: We need it to be based on the HIR to reject cases like `(T): ?Sized` at which point AST lowering remains a good enough place)

bors commented:
:umbrella: The latest upstream changes (presumably #143173) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16011625717/job/45170478002) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45170478002) [(plain)](https://github.com/rust-lang/rust/commit/008507c0cdae597cef0735d7f2dce0a56ea8f424/checks/45170478002/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests/ui/sized-hierarchy/pointee-validation.rs stdout ----
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sized-hierarchy/pointee-validation/pointee-validation.stderr`
diff of stderr:

9    = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
10 
11 error[E0658]: const trait impls are experimental
-   --> $DIR/pointee-validation.rs:12:53
+   --> $DIR/pointee-validation.rs:12:55
13    |
14 LL | const fn test<T, U>() where T: const PointeeSized, U: [const] PointeeSized {}
-    |                                                     ^^^^^^^^^
+    |                                                       ^^^^^^^
16    |
17    = note: see issue #67792 <https://github.com/rust-lang/rust/issues/67792> for more information
18    = help: add `#![feature(const_trait_impl)]` to the crate attributes to enable

42   --> $SRC_DIR/core/src/marker.rs:LL:COL
43 
44 error: `[const]` can only be applied to `#[const_trait]` traits
-   --> $DIR/pointee-validation.rs:12:53
+   --> $DIR/pointee-validation.rs:12:55
46    |
47 LL | const fn test<T, U>() where T: const PointeeSized, U: [const] PointeeSized {}
-    |                                                     ^^^^^^^^^ can't be applied to `PointeeSized`
+    |                                                       ^^^^^^^ can't be applied to `PointeeSized`
49    |
50 note: `PointeeSized` can't be used with `[const]` because it isn't annotated with `#[const_trait]`
51   --> $SRC_DIR/core/src/marker.rs:LL:COL

61    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
62 
63 error: `[const]` can only be applied to `#[const_trait]` traits
-   --> $DIR/pointee-validation.rs:12:53
+   --> $DIR/pointee-validation.rs:12:55
65    |
66 LL | const fn test<T, U>() where T: const PointeeSized, U: [const] PointeeSized {}
-    |                                                     ^^^^^^^^^ can't be applied to `PointeeSized`
+    |                                                       ^^^^^^^ can't be applied to `PointeeSized`
68    |
69 note: `PointeeSized` can't be used with `[const]` because it isn't annotated with `#[const_trait]`
70   --> $SRC_DIR/core/src/marker.rs:LL:COL

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/sized-hierarchy/pointee-validation.rs:12:55
-   --> /checkout/tests/ui/sized-hierarchy/pointee-validation.rs:12:55
-   --> /checkout/tests/ui/sized-hierarchy/pointee-validation.rs:12:55
---
To only update this specific test, also pass `--test-args sized-hierarchy/pointee-validation.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/sized-hierarchy/pointee-validation.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/sized-hierarchy/pointee-validation" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0658]: const trait impls are experimental
##[error]  --> /checkout/tests/ui/sized-hierarchy/pointee-validation.rs:12:32
   |
LL | const fn test<T, U>() where T: const PointeeSized, U: [const] PointeeSized {}
   |                                ^^^^^
   |
   = note: see issue #67792 <https://github.com/rust-lang/rust/issues/67792> for more information
   = help: add `#![feature(const_trait_impl)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: const trait impls are experimental
##[error]  --> /checkout/tests/ui/sized-hierarchy/pointee-validation.rs:12:55
   |
LL | const fn test<T, U>() where T: const PointeeSized, U: [const] PointeeSized {}
   |                                                       ^^^^^^^
   |
   = note: see issue #67792 <https://github.com/rust-lang/rust/issues/67792> for more information
   = help: add `#![feature(const_trait_impl)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0107]: trait takes 0 generic arguments but 1 generic argument was supplied
##[error]  --> /checkout/tests/ui/sized-hierarchy/pointee-validation.rs:8:20
   |
LL | struct T where (): PointeeSized<(), Undefined = ()>;
   |                    ^^^^^^^^^^^^-------------------- help: remove the unnecessary generics
   |                    |
   |                    expected 0 generic arguments

error[E0220]: associated type `Undefined` not found for `PointeeSized`
##[error]  --> /checkout/tests/ui/sized-hierarchy/pointee-validation.rs:8:37
   |
LL | struct T where (): PointeeSized<(), Undefined = ()>;
   |                                     ^^^^^^^^^ associated type `Undefined` not found

error: `const` can only be applied to `#[const_trait]` traits
##[error]  --> /checkout/tests/ui/sized-hierarchy/pointee-validation.rs:12:32
   |
LL | const fn test<T, U>() where T: const PointeeSized, U: [const] PointeeSized {}
   |                                ^^^^^ can't be applied to `PointeeSized`
   |
note: `PointeeSized` can't be used with `const` because it isn't annotated with `#[const_trait]`
  --> /rustc/FAKE_PREFIX/library/core/src/marker.rs:192:1

error: `[const]` can only be applied to `#[const_trait]` traits
##[error]  --> /checkout/tests/ui/sized-hierarchy/pointee-validation.rs:12:55
   |
LL | const fn test<T, U>() where T: const PointeeSized, U: [const] PointeeSized {}
   |                                                       ^^^^^^^ can't be applied to `PointeeSized`
   |
note: `PointeeSized` can't be used with `[const]` because it isn't annotated with `#[const_trait]`
  --> /rustc/FAKE_PREFIX/library/core/src/marker.rs:192:1

error: `const` can only be applied to `#[const_trait]` traits
##[error]  --> /checkout/tests/ui/sized-hierarchy/pointee-validation.rs:12:32
   |
LL | const fn test<T, U>() where T: const PointeeSized, U: [const] PointeeSized {}
   |                                ^^^^^ can't be applied to `PointeeSized`
   |
note: `PointeeSized` can't be used with `const` because it isn't annotated with `#[const_trait]`
  --> /rustc/FAKE_PREFIX/library/core/src/marker.rs:192:1
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

error: `[const]` can only be applied to `#[const_trait]` traits
##[error]  --> /checkout/tests/ui/sized-hierarchy/pointee-validation.rs:12:55
   |
LL | const fn test<T, U>() where T: const PointeeSized, U: [const] PointeeSized {}
   |                                                       ^^^^^^^ can't be applied to `PointeeSized`
   |
note: `PointeeSized` can't be used with `[const]` because it isn't annotated with `#[const_trait]`
  --> /rustc/FAKE_PREFIX/library/core/src/marker.rs:192:1
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

error: aborting due to 8 previous errors

Some errors have detailed explanations: E0107, E0220, E0658.
For more information about an error, try `rustc --explain E0107`.
---
diff of stderr:

38    | ++++++++++++++
39 
40 error: `[const]` can only be applied to `#[const_trait]` traits
-   --> $DIR/const-trait-bounds-trait-objects.rs:18:19
+   --> $DIR/const-trait-bounds-trait-objects.rs:18:23
42    |
43 LL | const fn take(_: &dyn [const] NonConst) {}
-    |                   ^^^^^^^^^^^ can't be applied to `NonConst`
+    |                       ^^^^^^^ can't be applied to `NonConst`
45    |
46 help: mark `NonConst` as `#[const_trait]` to allow it to have `const` implementations
47    |

Note: some mismatched output was normalized before being compared
---
To only update this specific test, also pass `--test-args traits/const-traits/const-trait-bounds-trait-objects.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/traits/const-traits/const-trait-bounds-trait-objects.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/traits/const-traits/const-trait-bounds-trait-objects" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2021"
stdout: none
--- stderr -------------------------------
error: const trait bounds are not allowed in trait object types
##[error]  --> /checkout/tests/ui/traits/const-traits/const-trait-bounds-trait-objects.rs:9:17
   |
LL |     let _: &dyn const Trait; //~ ERROR const trait bounds are not allowed in trait object types
   |                 ^^^^^^^^^^^

error: `[const]` is not allowed here
##[error]  --> /checkout/tests/ui/traits/const-traits/const-trait-bounds-trait-objects.rs:10:17
   |
LL |     let _: &dyn [const] Trait; //~ ERROR `[const]` is not allowed here
   |                 ^^^^^^^
   |
   = note: trait objects cannot have `[const]` trait bounds

error: const trait bounds are not allowed in trait object types
##[error]  --> /checkout/tests/ui/traits/const-traits/const-trait-bounds-trait-objects.rs:15:25
   |
LL | const fn handle(_: &dyn const NonConst) {}
   |                         ^^^^^^^^^^^^^^

error: `[const]` is not allowed here
##[error]  --> /checkout/tests/ui/traits/const-traits/const-trait-bounds-trait-objects.rs:18:23
   |
LL | const fn take(_: &dyn [const] NonConst) {}
   |                       ^^^^^^^
   |
   = note: trait objects cannot have `[const]` trait bounds

error: `const` can only be applied to `#[const_trait]` traits
##[error]  --> /checkout/tests/ui/traits/const-traits/const-trait-bounds-trait-objects.rs:15:25
   |
LL | const fn handle(_: &dyn const NonConst) {}
   |                         ^^^^^ can't be applied to `NonConst`
   |
help: mark `NonConst` as `#[const_trait]` to allow it to have `const` implementations
   |
LL | #[const_trait] trait NonConst {}
   | ++++++++++++++

error: `[const]` can only be applied to `#[const_trait]` traits
##[error]  --> /checkout/tests/ui/traits/const-traits/const-trait-bounds-trait-objects.rs:18:23
   |
LL | const fn take(_: &dyn [const] NonConst) {}
   |                       ^^^^^^^ can't be applied to `NonConst`
   |
help: mark `NonConst` as `#[const_trait]` to allow it to have `const` implementations
   |
LL | #[const_trait] trait NonConst {}
   | ++++++++++++++
---
diff of stderr:

43    |                            ^
44 
45 error: `[const]` can only be applied to `#[const_trait]` traits
-   --> $DIR/mutually-exclusive-trait-bound-modifiers.rs:3:29
+   --> $DIR/mutually-exclusive-trait-bound-modifiers.rs:3:31
47    |
48 LL | const fn maybe_const_maybe<T: [const] ?Sized>() {}
-    |                             ^^^^^^^^^ can't be applied to `Sized`
+    |                               ^^^^^^^ can't be applied to `Sized`
50    |
51 note: `Sized` can't be used with `[const]` because it isn't annotated with `#[const_trait]`
52   --> $SRC_DIR/core/src/marker.rs:LL:COL

53 
54 error: `[const]` can only be applied to `#[const_trait]` traits
-   --> $DIR/mutually-exclusive-trait-bound-modifiers.rs:3:29
+   --> $DIR/mutually-exclusive-trait-bound-modifiers.rs:3:31
56    |
57 LL | const fn maybe_const_maybe<T: [const] ?Sized>() {}
-    |                             ^^^^^^^^^ can't be applied to `Sized`
+    |                               ^^^^^^^ can't be applied to `Sized`
59    |
60 note: `Sized` can't be used with `[const]` because it isn't annotated with `#[const_trait]`
61   --> $SRC_DIR/core/src/marker.rs:LL:COL

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs:3:31
-   --> /checkout/tests/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs:3:31
+   --> $DIR/mutually-exclusive-trait-bound-modifiers.rs:3:31
---
To only update this specific test, also pass `--test-args traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error: `[const]` trait not allowed with `?` trait polarity modifier
##[error]  --> /checkout/tests/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs:3:39
   |
LL | const fn maybe_const_maybe<T: [const] ?Sized>() {}
   |                               ------- ^
   |                               |
   |                               there is not a well-defined meaning for a `[const] ?` trait

error: `const` trait not allowed with `?` trait polarity modifier
##[error]  --> /checkout/tests/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs:8:25
   |
LL | fn const_maybe<T: const ?Sized>() {}
   |                   ----- ^
   |                   |
   |                   there is not a well-defined meaning for a `const ?` trait

error: `[const]` trait not allowed with `!` trait polarity modifier
##[error]  --> /checkout/tests/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs:12:42
   |
LL | const fn maybe_const_negative<T: [const] !Trait>() {}
   |                                  ------- ^
   |                                  |
   |                                  there is not a well-defined meaning for a `[const] !` trait

error: `const` trait not allowed with `!` trait polarity modifier
##[error]  --> /checkout/tests/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs:16:28
   |
LL | fn const_negative<T: const !Trait>() {}
   |                      ----- ^
   |                      |
   |                      there is not a well-defined meaning for a `const !` trait

error: negative bounds are not supported
##[error]  --> /checkout/tests/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs:12:42
   |
LL | const fn maybe_const_negative<T: [const] !Trait>() {}
   |                                          ^

error: negative bounds are not supported
##[error]  --> /checkout/tests/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs:16:28
   |
LL | fn const_negative<T: const !Trait>() {}
   |                            ^

error: `[const]` can only be applied to `#[const_trait]` traits
##[error]  --> /checkout/tests/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs:3:31
   |
LL | const fn maybe_const_maybe<T: [const] ?Sized>() {}
   |                               ^^^^^^^ can't be applied to `Sized`
   |
note: `Sized` can't be used with `[const]` because it isn't annotated with `#[const_trait]`
  --> /rustc/FAKE_PREFIX/library/core/src/marker.rs:158:1

error: `[const]` can only be applied to `#[const_trait]` traits
##[error]  --> /checkout/tests/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs:3:31
   |
LL | const fn maybe_const_maybe<T: [const] ?Sized>() {}
   |                               ^^^^^^^ can't be applied to `Sized`
   |
note: `Sized` can't be used with `[const]` because it isn't annotated with `#[const_trait]`
  --> /rustc/FAKE_PREFIX/library/core/src/marker.rs:158:1
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

error: `const` can only be applied to `#[const_trait]` traits
##[error]  --> /checkout/tests/ui/traits/const-traits/mutually-exclusive-trait-bound-modifiers.rs:8:19
   |
LL | fn const_maybe<T: const ?Sized>() {}
   |                   ^^^^^ can't be applied to `Sized`
   |
note: `Sized` can't be used with `const` because it isn't annotated with `#[const_trait]`
  --> /rustc/FAKE_PREFIX/library/core/src/marker.rs:158:1

error: aborting due to 9 previous errors
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #143958) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors reviewed (approved):
Needs a rebase, r=me afterwards. Thanks for the thoughtful comments as well.

compiler-errors commented:
@rustbot author

fmease committed:
Don't reject *multiple* relaxed bounds, reject *duplicate* ones.

Having multiple relaxed bounds like `?Sized + ?Iterator` is actually *fine*.
We actually want to reject *duplicate* relaxed bounds like `?Sized + ?Sized`
because these most certainly represent a user error.

Note that this doesn't mean that we accept more code because a bound like
`?Iterator` is still invalid as it's not relaxing a *default* trait and
the only way to define / use more default bounds is under the experimental
and internal feature `more_maybe_bounds` plus `lang_items` plus unstable
flag `-Zexperimental-default-bounds` (historical context: for the longest
time, bounds like `?Iterator` were actually allowed and lead to a hard
warning).

Ultimately, this simply *reframes* the diagnostic. The scope of
`more_maybe_bounds` / `-Zexperimental-default-bounds` remains unchanged
as well.

fmease commented:
(lol I should've just built the compiler for the rebase instead of doing a CI-directed one, that was painful, thanks tired me doing a night shift)

Rebased, added some more (doc) comments and got rid of [the `more_maybe_bounds` advertisement](https://github.com/rust-lang/rust/pull/142693#discussion_r2211443889) (didn't do it as a follow-up so I can already use this as a good base branch for #135331.

@bors r=compiler-errors

bors commented:
:pushpin: Commit 82a02aefe07092c737c852daccebf49ca25507e3 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 82a02aefe07092c737c852daccebf49ca25507e3 -->
<!-- homu: {"type":"Approved","sha":"82a02aefe07092c737c852daccebf49ca25507e3","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

fmease commented:
@bors rollup- alleviate pressure
